src/main/scala/com/yahoo/tensorflowonspark/TFModel.scala

```diff
@@ -0,0 +1,348 @@
+              val floatArr = Array.ofDim[Float](numElements)
+              floatBuf.get(floatArr, 0, numElements)
+              floatArr
+            case DataType.INT32 =>
+              val intBuf = buf.asInstanceOf[IntBuffer]
+              val intArr = Array.ofDim[Int](numElements)
+              intBuf.get(intArr, 0, numElements)
+              intArr
+            case DataType.INT64 =>
+              val longBuf = buf.asInstanceOf[LongBuffer]
+              val longArr = Array.ofDim[Long](numElements)
+              longBuf.get(longArr, 0, numElements)
+              longArr
+            case DataType.STRING =>
+              val byteBuf = buf.asInstanceOf[ByteBuffer]
+              val byteArr = Array.ofDim[Byte](buf.remaining())
+              byteBuf.get(byteArr)
+            case DataType.UINT8 =>
+              val byteBuf = buf.asInstanceOf[ByteBuffer]
+              val byteArr = Array.ofDim[Byte](buf.remaining())
+              byteBuf.get(byteArr)
+            case unsupportedType =>
+              Array()
+          }
+        }
+      }
+      result += Row(slices:_*)
+    }
+    result.toList
+  }
+
+  /**
+    * Transforms a Dataset of input data into an output DataFrame of predictions.
+    * Note: input columns are dropped during transform.
+    */
+  override def transform(dataset: Dataset[_]): DataFrame = {
+    val spark = dataset.sparkSession
+
+    val inputColumns = this.getInputMapping.keys.toSeq
+    val inputTensorNames = this.getInputMapping.values
+    val outputTensorNames = this.getOutputMapping.keys.toSeq
+
+    val inputDF = dataset.select(inputColumns.head, inputColumns.tail: _*)
+    val inputSchema = inputDF.schema
+    val outputSchema = transformSchema(inputSchema)
+
+    val outputRDD = inputDF.rdd.mapPartitions { iter: Iterator[Row] =>
+      if (TFModel.model == null || TFModel.modelDir != this.getModel) {
+        // load model into a per-executor singleton reference, if needed.
+        TFModel.modelDir = this.getModel
+        TFModel.model = SavedModelBundle.load(this.getModel, this.getTag)
+        TFModel.graph = TFModel.model.graph
+        TFModel.sess = TFModel.model.session
+      }
+
+      var results = List.empty[Row]
+      val groupedIter = iter.grouped(this.getBatchSize)
+      while (groupedIter.hasNext) {
+        val batch = groupedIter.next()
+
+        // get input batch of Rows and convert to list of input Tensors
+        val inputTensors = batch2tensors(batch, inputSchema)
+
+        var runner = TFModel.sess.runner()
+
+        // feed input tensors
+        for ((name, tensor) <- inputTensors) {
+          runner = runner.feed(this.getInputMapping(name), tensor)
+        }
+        // fetch output tensors
+        for (name <- outputTensorNames) {
+          runner = runner.fetch(name)
+        }
+
+        // run the graph
+        val outputTensors = runner.run()
+
+        assert(outputTensors.map(_.shape).map(s => if (s.isEmpty) 0L else s.apply(0)).distinct.size == 1,
```

Comment:
What could cause this assertion to fail?

Reply:
If the user selects output tensors which aren't the same shape, e.g. a per-input-record output + some per-batch output.